### ISSUE TITLE
BayesianNetwork.from_samples, model.predict(test) not making prediction in all categories

### ISSUE THREAD

**Issue by Geldren1** [jmschrei/pomegranate#425]
Hello.
Python = 3.6.0
Pomegranate = 0.9.0

I am using a BayesianNetwork.from_samples on training data, then asking it to make predictions on test data. There are a potential 4 categories of the target variable and there are examples of each in both the test and training data. When I use the model.predict (test) I get predictions for 3/4 categories. There is always a category missing.  I have tried using psuedocount (values 1 & 2) but this does not help. Any ideas?

Many thanks in advance. Loving Pomegranate.

**Comment by jmschrei** [comment]
Howdy! Can you describe what you mean a bit more? Do you mean that there are four variables and it's only making predictions for three, or that there is one variable that can take four values but it never predicts the fourth one? How are you building the Bayes net? What does the distribution look like after training?

**Comment by Geldren1** [comment]
Hello! Many thanks for your rapid response.
I am talking about the later case (one variable that can take four discrete values but never predicts for 1 of them). More specifically, the variable can take the values 1, 2, 3, 4. The predict function will make predictions for 3 of the 4 options. The options that tend to get missed out are 1 and 4 but it does change with different iterations. I checked that there were examples of all 4 categories in both the training and test sets. There were. 
I was experimenting with it a bit more and I found if I changed the proportion that I split the data into training and test sets, I could increase the probability that I would get a model for the from_samples function that predicts across all categories. Specifically if the TEST set was large (90% of the total data available) and run the training multiple independent times, I would get a model that predicted across all categories within about 15 iterations. 
I thought it might be something to do with the Training Set. There are about 12 variables and each can also take the values 1 to 4. it is sparse. I think the resultant model struggles if it hasn't seen a particular combination of input variables. But if this were the case then surely a bigger training set would perform better? Or perhaps the model will always come up with a prediction even if the confidence in that prediction is very low? 
Your thoughts based on your knowledge of the inner workings of pomegranate would be very much appreciated!
Helen

    On ‎Tuesday‎, ‎1‎ ‎May‎ ‎2018‎ ‎17‎:‎34‎:‎20‎ ‎BST, Jacob Schreiber <notifications@github.com> wrote:  
 
 
Howdy! Can you describe what you mean a bit more? Do you mean that there are four variables and it's only making predictions for three, or that there is one variable that can take four values but it never predicts the fourth one? How are you building the Bayes net? What does the distribution look like after training?

**Comment by jmschrei** [comment]
I think that @Wesserg it correct. It's difficult for it to predict things that rarely occur. You can try using `predict_proba` instead of `predict` to see the probability of the rare classes instead of the hard prediction. Is it possible that the rare conditions that lead to a prediction of 1 don't occur in the test set naturally? Can you expand on the number of samples you're training on and the structure of the Bayesian network? Having a feature space of 4**12 will be difficult one way or another.

**Comment by jmschrei** [comment]
Please re-open if you have further comments or are still encountering the issue.
